### DISCUSSION TITLE
Opening a .dig file with doubleclick or command line

### DISCUSSION THREAD

**Discussion by TXAE** [hneemann/Digital#916] [Q&A]
I noticed it is not possible to open .dig-files on Win10 with Digital by directly by associating the .dig-extension with Digital.exe. This simply launches naked Digital.exe without opening the double-clicked .dig-file. Drag and Drop of a .dig-file onto Digital GUI also not possible.
Are these core functions planned for a future release?

I then tried to open a .dig-file via command line (with the intention to later create a .bat file to be able to launch Digital with a circuit opened in it by double clicking this one .bat-file). I moved BillieJean.dig to the same folder as Digital.jar and tried to open the circuit with
`java -jar Digital.jar BillieJean.dig`

I noticed this opens up Digital and even names the Window "BillieJean.dig - Digital" as if the circuit was open, but there was only naked Digital. This is most definitely a bug since it suggests Digital thinks the circuit is open while it in fact isn't.
I then tried entering passing the full path as an argument since I noticed that Digital windows are always named after the full path, e.g. "C:\Users\...\BillieJean.dig - Digital"
So I tried:
`java -jar Digital.jar C:\Users\...\BillieJean.dig`

(... is shortened, I did not really type in ... there)

This launched a Digital window with again a blank screen, but also a significantly shorter path compared to the path I entered. So the path must be cut off somewhere I thought. Alright, moved BillieJean.dig to C:\ and tried again:
`java -jar Digital.jar C:\BillieJean.dig`

Again, no luck, blank screen, but Digital window named as if the file is open. 

The only way to open .dig-files seems to be to launch Digital first and then open the file from within Digital, requiring at least 2 steps (if file recently opened), more steps if the file that hasn't been opened previously. Any tips/ideas?

**Comment by hneemann** [comment]
Associating the .dig file extension with the simulator is possible without any problems. 
I use this all the time to open *.dig files with a double click.
I suspect that there is some problem on your machine.
Have any error reports been generated in your home folder? (File name similar to `Digital_2021-04-20_21-55-51.log`)
If so, could you attach these files to this ticket?

**Reply by TXAE** [reply]
Good to know! I tried on my work and my private laptop (both Win10) and it doesn't work on either of them. Digital and BillieJean.dig are stock from this repo (no changes). Same behaviour with both Digital.exe and Digital_noD3D.exe
JRE 17.0.2
VM 17.0.2+8-LTS-86
Build git-Revision: v0.28
No such .log file is created on either laptop. No other failure indication either. There's simply a blank screen named after the file I was trying to open.
Looks like this on both laptops for BillieJean.dig, but I tried other .dig-files too with the same outcome:

![image](https://user-images.githubusercontent.com/70020564/153001099-37dcbf63-ed02-4dc3-81e1-ba7c5b338e22.png)

Any ideas about the problems with command line?

**Reply by hneemann** [reply]
The file name is recognized and processed correctly, otherwise the file name would appear in the window title.
Is it a display problem? Can you click on the "play button", and does an audio output sound?

**Reply by TXAE** [reply]
When double-clicking I can click "play button", but no sound. Behaves exactly as an empty circuit would.

Sound plays on both machines when opening BillieJean.dig from within Digital (behaves as expected in this case).

Trying on a 100% fresh VM...

**Comment by TXAE** [comment]
No luck with the fresh Win10 VM either. Also tried the latest pre-release, no change. I have tried on 3 different systems now and the association doesn't work on any of them :( Did you try on another system?

Another question: Can you explain this driver behavior? Shouldn't out be always High-Z?
Noticed this on a bus first. The driver would put seemingly random stuff on the bus despite in remaining in High-Z. The attached circuit is a simpler reproduction of what might be a bug when driver sel is connected to a clock signal.

[Driver_Bug.zip](https://github.com/hneemann/Digital/files/8025845/Driver_Bug.zip)

**Reply by hneemann** [reply]
That's not a bug, it's intentional. 
When high-z outputs are connected to an input, the circuit behaves similar to a real CMOS gate. The read value is more or less random. Maybe this could be adjusted if, as in this example, no real gate is connected, but only an output pin. (Output pins behave in the simulation like normal inputs of logic gates). 
See also #665 and the references in #36.

**Reply by TXAE** [reply]
Thanks for the great explanation! Maybe include random output for High-Z input in help-text?

Regarding the main problem (opening with doubleclick): I've tested on other students laptops now and even asked your colleague from Mannheim, Dr. Ing. Bernhard Spitzer, about this. All of them have to open .dig files from within Digital. The association doesn't work in a single case but yours. Please verify this on an independent machine.

**Reply by hneemann** [reply]
@TXAE 
If the simulator is started without passing a file name, does it open the file that was last used, or does the simulator open with an empty circuit area?
And if it opens with an empty circuit area, are there file names in the menu "File->Open Recent"?

**Reply by TXAE** [reply]
Simulator opens with empty circuit area in that case.
Yes, there are (multiple) file names in the recent-menu.

Thanks for looking into this @hneemann
Appreciate it!

**Reply by hneemann** [reply]
@TXAE 
When the simulator opens with an empty circuit panel, does the last used filename show up in the window title, as seems to be the case when a filename is passed at startup?

**Reply by TXAE** [reply]
@hneemann 
A) Digital.exe launched without passing file: no filename in window title, window title just Digital in this case
B) Passing file (i.e. double-clicking associated .dig-file): Window title is full path, but empty circuit. Quick & dirty video:

https://user-images.githubusercontent.com/70020564/153554640-8953c0a9-bac0-4ac8-9cda-000997f0e68c.mp4

**Reply by hneemann** [reply]
Please download the latest [pre-release](https://infdigital.dhbw-mosbach.de/) version and start the simulator with the following command (replace the last part of the command with a file that exists on your computer)

```
java -Dorg.slf4j.simpleLogger.defaultLogLevel=debug -jar Digital.jar ~/temp/Digital/dig/z.dig
```
Please attache the console output to this discussion.

**Reply by TXAE** [reply]
Again, same behavior: Window title = full file path, but empty circuit.

C:\Users\sveni\Downloads\Digital_v0.28-56-g9cce4fb\Digital>java -Dorg.slf4j.simpleLogger.defaultLogLevel=debug -jar Digital.jar C:\Users\sveni\Downloads\Digital_v0.28-56-g9cce4fb\Digital\examples\misc\BillieJean.dig
Use default settings!
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibraryFolder - found 126 files in C:\Users\sveni\Downloads\Digital_v0.28-56-g9cce4fb\Digital\lib
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener CircuitComponent, listeners: 1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener InsertHistory, listeners: 2
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener LibrarySelector, listeners: 3

This was added on close of the Digital instance:
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener LibrarySelector, listeners: 2
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener InsertHistory, listeners: 1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener CircuitComponent, listeners: 0

**Reply by hneemann** [reply]
@TXAE 
Please try again with [v0.28-57-ga3b178c](https://infdigital.dhbw-mosbach.de/)

Does it always say: "Use default settings!"?

**Reply by TXAE** [reply]
Yes, but some changes with [v0.28-57-ga3b178c](https://infdigital.dhbw-mosbach.de/)

C:\Users\sveni\Downloads\Digital_v0.28-57-ga3b178c\Digital>java -Dorg.slf4j.simpleLogger.defaultLogLevel=debug -jar Digital.jar C:\Users\sveni\Downloads\Digital_v0.28-56-g9cce4fb\Digital\examples\misc\BillieJean.dig

Use default settings!
**[main] INFO de.neemann.digital.gui.Main - v0.28-57-ga3b178c**
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibraryFolder - found 126 files in C:\Users\sveni\Downloads\Digital_v0.28-57-ga3b178c\Digital\lib
**[AWT-EventQueue-0] DEBUG de.neemann.digital.gui.Main - create with given circuit**
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener CircuitComponent, listeners: 1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener InsertHistory, listeners: 2
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener LibrarySelector, listeners: 3

Same behavior, same output on close as before.

**Reply by hneemann** [reply]
@TXAE 
Please try again with [v0.28-59-ga9a3657](https://infdigital.dhbw-mosbach.de/).

It seems to me that Digital always opens in "tutorial mode".

**Reply by hneemann** [reply]
@TXAE 
The latest pre-release [v0.28-60-g1cb8df1](https://infdigital.dhbw-mosbach.de/) skips the tutorial if a file name is passed in at startup.
Maybe this modification will fix this problem.

**Reply by TXAE** [reply]
A success! At least partial. Opening with provided console command does indeed show the BillieJean.dig circuit. 
Associating .dig-files with [v0.28-60-g1cb8df1](https://infdigital.dhbw-mosbach.de/) still shows a blank circuit. I made sure the association is with [v0.28-60-g1cb8df1](https://infdigital.dhbw-mosbach.de/) by renaming "Digital.exe" to "Digital_v0.28-60.exe" bc I have several Digital versions now. "Opens with" confirms Windows is opening .dig files with "Digital_v0.28-60.exe". Still, no circuit when double clicking :(
==> Only the commands passed with cmd make Digital behave as expected and opens the circuit
==> Opening with double-click still shows blank screen with Window-title = full file path.

Another observation I made bc you mentioned tutorial mode: "View>Start Tutorial" doesn't do anything. When a circuit is loaded, it makes the circuit disappear. When there's a blank page, seemingly nothing happens. I tested this on 2 machines.

Console output with successful opening of BillieJeans.dig:

C:\Users\sveni\Downloads\Digital_v0.28-60-g1cb8df1\Digital>java -Dorg.slf4j.simpleLogger.defaultLogLevel=debug -jar Digital.jar C:\Users\svenia\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc\BillieJean.dig

[main] DEBUG de.neemann.digital.gui.Main - v0.28-60-g1cb8df1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibraryFolder - found 126 files in C:\Users\svenia\Downloads\Digital_v0.28-60-g1cb8df1\Digital\lib
[AWT-EventQueue-0] DEBUG de.neemann.digital.gui.Main - create with given file C:\Users\svenia\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc\BillieJean.dig
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener CircuitComponent, listeners: 1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener InsertHistory, listeners: 2
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener LibrarySelector, listeners: 3
[AWT-EventQueue-0] DEBUG de.neemann.digital.gui.Main - loadFile: C:\Users\sveni\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc\BillieJean.dig
[AWT-EventQueue-0] DEBUG de.neemann.digital.gui.Main - set library root: C:\Users\sveni\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc\BillieJean.dig
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - rescan folder
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibraryFolder - found 5 files in C:\Users\sveni\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.elements.Circuit - load C:\Users\sveni\Downloads\Digital_v0.28-60-g1cb8df1\Digital\examples\misc\BillieJean.dig
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.elements.Circuit - load stream
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - added library listener LibraryTreeModel, listeners: 4

On close:
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener LibrarySelector, listeners: 3
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener InsertHistory, listeners: 2
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener CircuitComponent, listeners: 1
[AWT-EventQueue-0] DEBUG de.neemann.digital.draw.library.ElementLibrary - removed library listener LibraryTreeModel, listeners: 0

**Reply by TXAE** [reply]
Nevermind, I don't know what changed, but even double clicking works now! 😍

**Reply by TXAE** [reply]
I hope the mentioned change "skips tutorial if a file name is passed in at startup" makes it to next production version! Would be great. Although this doesn't address the seemingly broken Tutorial, but simply skips it.

New observation (minor): Changing "View>Component Tree View (F5)" doesn't save. It always shows for me now.

**Reply by TXAE** [reply]
Just tried on another student's laptop. Opening a circuit and then clicking "View>Start Tutorial" makes the circuit disappear in his case too. Tutorial seems to be broken. And he's on v0.24!

**Reply by hneemann** [reply]
@TXAE 
Could you please try again with the simulator window in non-maximum size? So that all around some 100 pixel space is available?

**Reply by TXAE** [reply]
This was it! Tutorial works if Digital window isn't maximized.

**Reply by hneemann** [reply]
@TXAE 
That's interesting: On Linux/Gnome this does not happen. The tutorial dialog is always visible.

**Reply by TXAE** [reply]
Yeah prob Windows is opening the little Tutorial window to the left next to the maximized window invisible to the user, resulting in it looking like what I've described previously: Simply going to a blank screen. As the Tutorial window isn't its own instance, one can't even see the tutorial window by alt-tabbing.
Maybe change this so that tutorial window is always on top of Digital window (e.g. top left?) and maybe its own instance so that its visible by alt-tabbing?

This doesn't explain why it would always try to open the tutorial on startup even when I passed a file. I think it would be better to find out what causes opening tutorial mode on startup instead of skipping tutorial mode when a file is passed.

**Reply by hneemann** [reply]
@TXAE 
Please try to show the tutoral dialog in max window size using [v0.28-61-gcf5441a](https://infdigital.dhbw-mosbach.de/).

**Reply by hneemann** [reply]
@TXAE 
> This doesn't explain why it would always try to open the tutorial on startup even when I passed a file. I think it would be better to find out what causes opening tutorial mode on startup instead of skipping tutorial mode when a file is passed.

Yes it does! The "Tutorial Mode" is always activated at startup until the tutorial is either completed or skipped. Neither is possible if the dialog is not visible to the user.

What else would be possible is that the settings file ".digital.cfg" could not be written, because in this file the status of the tutorial is stored. That would be then some "security feature" of the operating system.

**Reply by TXAE** [reply]
@hneemann perfect! Tutorial now shows on top left on my Win10:

![image](https://user-images.githubusercontent.com/70020564/153583034-5158ddcd-ff46-4cf9-9657-55c373309e8b.png)

**Reply by TXAE** [reply]
> Yes it does! The "Tutorial Mode" is always activated at startup until the tutorial is either completed or skipped. Neither is possible if the dialog is not visible to the user.
> 
> What else would be possible is that the settings file ".digital.cfg" could not be written, because in this file the status of the tutorial is stored. That would be then some "security feature" of the operating system.

I see! Ty for the great explanation, again! It was definitely the first problem (tutorial not visible to user) as once I saw the tutorial and was able to skip it, it never showed again, so writing the settings file ".digital.cfg" worked without a problem.

**Reply by TXAE** [reply]
Just noticed that Tunnel show state since one of the latest pre-releases! Much appreciated!

**Comment by xotmatrix** [comment]
I can verify that this works for me on Windows 10 but I've had Digital installed for years and do not known if something has changed with the way .dig files are associated since then.

**Reply by TXAE** [reply]
Interesting, thanks for sharing!
